peps/pep-0747.rst

```diff
@@ -0,0 +1,903 @@
+   | | +-------------------------+  | | 
+   | | Annotation expression object | | 
+   | +------------------------------+ | 
+   | Object                           | = object         
+   +----------------------------------+ 
+
+
+Using TypeExprs
+---------------
+
+A ``TypeExpr`` type may be used in any context where a type annotation is
+valid, as a function parameter type, a return type, or a variable type:
+
+::
+
+   def is_union_type(typx: TypeExpr) -> bool: ...  # parameter type
+
+::
+
+   def union_of[S, T](s: TypeExpr[S], t: TypeExpr[T]) \
+       -> TypeExpr[S | T]: ...  # return type
+
+::
+
+   STR_TYPE: TypeExpr = str  # variable type
+   assert_type(STR_TYPE, TypeExpr[str])
+
+Note however that an *unannotated* variable assigned a type expression literal
+will not be inferred to be of ``TypeExpr`` type by type checkers because PEP
+484 :pep:`reserves that syntax for defining type aliases <484#type-aliases>`:
+
+-  No:
+
+   ::
+
+      STR_TYPE = str  # OOPS; treated as a type alias!
+
+If you want a type checker to recognize a type expression literal in a bare
+assignment you’ll need to explicitly declare the assignment-target as
+having ``TypeExpr`` type:
+
+-  Yes:
+
+   ::
+
+      STR_TYPE: TypeExpr = str
+
+-  Yes:
+
+   ::
+
+      STR_TYPE: TypeExpr
+      STR_TYPE = str
+
+-  Okay, but discouraged:
+
+   ::
+
+      STR_TYPE = str  # type: TypeExpr  # the type comment is significant
+
+``TypeExpr`` values can be passed around and assigned just like normal
+values:
+
+::
+
+   def swap1[S, T](t1: TypeExpr[S], t2: TypeExpr[T]) -> tuple[TypeExpr[T], TypeExpr[S]]:
+       t1_new: TypeExpr[T] = t2  # assigns a TypeExpr value to a new annotated variable
+       t2_new: TypeExpr[S] = t1
+       return (t1_new, t2_new)
+
+   def swap2[S, T](t1: TypeExpr[S], t2: TypeExpr[T]) -> tuple[TypeExpr[T], TypeExpr[S]]:
+       t1_new = t2  # assigns a TypeExpr value to a new unannotated variable
+       t2_new = t1
+       assert_type(t1_new, TypeExpr[T])
+       assert_type(t2_new, TypeExpr[S])
+       return (t1_new, t2_new)
+
+   # NOTE: A more straightforward implementation would use isinstance()
+   def ensure_int(value: object) -> None:
+       value_type: TypeExpr = type(value)  # assigns a type (a subtype of TypeExpr)
+       assert value_type == int
+
+
+TypeExpr Values
+---------------
+
+A variable of type ``TypeExpr[T]`` where ``T`` is a type, can hold any
+**type expression object** - the result of evaluating a `type expression`_
+at runtime - which is a subtype of ``T``.
+
+Incomplete expressions like a bare ``Optional`` or ``Union`` which do
+not spell a type are not ``TypeExpr`` values.
+
+``TypeExpr[...]`` is itself a ``TypeExpr`` value:
+
+::
+
+   OPTIONAL_INT_TYPE: TypeExpr = TypeExpr[Optional[int]]  # OK
+   assert isassignable(Optional[int], OPTIONAL_INT_TYPE)
+
+
+Stringified TypeExprs
+'''''''''''''''''''''
+
+A ``TypeExpr`` value may be a string literal that spells a forward
```

Comment:
It should be made clear that this section does not spell out a special case; it's a natural consequence of the definitions of "type expression" and TypeExpr.

Reply:
I rephrased as:

> A ``TypeExpr``, **as a consequence of being type expression**, may be a string literal that ...